### PR TITLE
feat: add sync gc-s3 command for S3 garbage collection

### LINKED ISSUE #226: Feature: Add `sync gc-s3` command for S3 garbage collection
## Problem

When `sync to-s3` backs up artifacts, it only uploads - it never deletes. This means S3 accumulates orphaned blobs over time:

1. Blob is tagged and backed up to S3
2. Tag is removed locally
3. Local GC deletes the untagged blob
4. `to-s3` syncs the updated manifest (tag removed) but doesn't delete the blob
5. Blob is now orphaned in S3

Over time, S3 can grow significantly larger than local storage. This becomes a problem during disaster recovery - restoring from S3 could fill the disk with orphaned blobs that were already GC'd locally.

### Why not use S3 lifecycle policies?

Lifecycle policies expire objects not touched in N days. But this creates a ticking clock that doesn't distinguish between:
- Orphaned garbage (should delete)
- Sync job silently broken (should NOT delete)
- Off-season, server idle for months (should NOT delete)

If the lifecycle period expires during downtime, the entire backup gets deleted.

## Proposed Solution

Add a new `gc-s3` subcommand that uses **S3's own metadata** as the source of truth:

```bash
magpie-ctl sync gc-s3 [--dry-run] [--execute]
```

### Algorithm

1. List all `.magpie` manifests in S3
2. Build set of all tagged hashes across all manifests
3. List all blobs in S3
4. Delete blobs not referenced by any manifest

### Key Design Decisions

**S3-only, no local state required:**
- Reads manifests from S3, not local storage
- Works even if the local server is down
- Self-consistent: S3's manifests define what S3 should keep
- Cannot accidentally wipe backup due to empty/partial local state

**Safe by default:**
- `--dry-run` is the default behavior
- Requires explicit `--execute` flag to actually delete
- Reports what would be deleted before any destructive action

**No server identity/UUID needed:**
- Since we're using S3's own manifests as source of truth, there's no risk of cross-server conflicts
- The manifests in S3 are the authoritative record of what's tagged

### Example Usage

```bash
# Preview what would be deleted
magpie-ctl sync gc-s3

# Actually delete orphaned blobs
magpie-ctl sync gc-s3 --execute

# With explicit bucket (if not in env)
MAGPIE_S3_BUCKET=my-bucket magpie-ctl sync gc-s3
```

### Expected Output

```
S3 Garbage Collection Preview:
  Manifests found: 42
  Tagged blobs: 156
  Total blobs in S3: 203
  Orphaned blobs: 47 (1.2 GB)

Would delete 47 blob(s). Run with --execute to proceed.
```

## Operational Guidance

This command should be run:
- Manually during active operational periods
- After confirming the local server is healthy and `to-s3` has run recently
- Never automated with a cron job (to avoid silent deletion during outages)

## Related

- `sync to-s3` - uploads tagged artifacts
- `sync from-s3` - restores from S3
- `gc` - local garbage collection

---

(AI-generated via Claude Code w/ Opus 4.5)

### PR DESCRIPTION
**Pull request by duplico**
## Summary

- Add new `gc-s3` subcommand to the `sync` command group for garbage collecting orphaned blobs from S3 backup storage
- The command uses S3's own manifests as the source of truth, allowing it to run independently of local storage state
- Safe by default: `--dry-run` is the default behavior, requiring explicit `--execute` flag to delete
- Supports both rclone and aws CLI, consistent with existing sync commands

## Implementation Details

The gc-s3 command:
1. Lists all `.magpie` manifest files in S3
2. Parses each manifest to build a set of all tagged blob hashes
3. Lists all blob files in S3 (`*/blobs/*`)
4. Identifies blobs not referenced by any manifest
5. Optionally deletes orphaned blobs (with `--execute`)

New helper functions added:
- `_list_s3_files_rclone` / `_list_s3_files_aws` - List files matching patterns
- `_get_s3_file_content_rclone` / `_get_s3_file_content_aws` - Read file content
- `_delete_s3_file_rclone` / `_delete_s3_file_aws` - Delete files
- `_get_s3_file_size_rclone` / `_get_s3_file_size_aws` - Get file sizes
- `_parse_manifest_content` - Parse manifest JSON for blob hashes
- `_extract_blob_name_from_path` - Extract blob name from S3 path
- `_find_orphaned_blobs` - Core algorithm to identify orphans

## Test Plan

- [x] Added 25 unit tests covering:
  - Manifest parsing edge cases (empty, null values, deduplication)
  - Blob path extraction
  - Orphan detection with mocked S3 operations
  - CLI behavior (dry-run default, execute mode, JSON output)
- [x] All existing unit tests pass (796 total)
- [x] Linting passes (`ruff check` and `ruff format`)

Closes #226

---

Generated with [Claude Code](https://claude.ai/code) w/ Opus 4.5

### CODE REVIEW

File: src/magpie/ctl/commands/sync.py [hunk 1]
```diff
@@ -854,3 +854,727 @@ def from_s3(
...
+        magpie-ctl sync gc-s3 --execute
+
+        # With explicit bucket
+        MAGPIE_S3_BUCKET=my-bucket magpie-ctl sync gc-s3
+    """
+    settings = ctx.settings
+
+    # If --execute is provided, disable dry_run
+    if execute:
+        dry_run = True  # Will be flipped to False below
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/ctl/commands/sync.py]: Fixed in commit afd48f9. Removed the redundant `dry_run = True` line so the logic now directly sets `dry_run = False` when `--execute` is provided.

(AI-generated via Claude Code w/ Opus 4.5)

File: src/magpie/ctl/commands/sync.py [hunk 2]
```diff
@@ -854,3 +854,727 @@ def from_s3(
...
+        get_content = _get_s3_file_content_aws
+
+    # Step 1: List all manifest files
+    if not quiet and not is_json_output():
+        click.echo("Scanning S3 for manifests...")
+
+    manifest_paths = list_files(bucket, prefix, "*.magpie", debug)
+    # Also try pattern that matches nested paths
+    if not manifest_paths:
+        manifest_paths = list_files(bucket, prefix, "**/.magpie", debug)
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/ctl/commands/sync.py]: Fixed in commit afd48f9. Changed to use `**/.magpie` pattern directly instead of trying `*.magpie` first. The `*.magpie` pattern wouldn't match the hidden `.magpie` files in artifact directories since they don't have a prefix before the dot.

(AI-generated via Claude Code w/ Opus 4.5)
